src/analysis.rs

```diff
@@ -27,358 +23,830 @@ pub fn generate_blackman(size: usize) -> impl Iterator<Item = f32> {
+        });
+
+        write_index += len;
+
+        if write_index >= RING_BUFFER_SIZE {
+            write_index -= RING_BUFFER_SIZE;
         }
 
-        false
+        self.write_index.store(write_index, Ordering::SeqCst);
     }
 
-    /// Read out the ring buffer (max `fft_size` samples)
-    fn get_float_time(&self, buffer: &mut [f32], fft_size: usize) {
-        // buffer is never empty when this call is made
-        debug_assert!(!self.buffer.is_empty());
-
-        // get a reference to the 'silence buffer'
-        let silence = self.buffer[0].silence();
+    pub fn read(&self, dst: &mut [f32], max_len: usize) {
+        let write_index = self.write_index.load(Ordering::SeqCst);
+        // let fft_size = self.fft_size.load(Ordering::SeqCst);
+        let len = dst.len().min(max_len);
+
+        dst.iter_mut()
+            .take(len)
+            .enumerate()
+            .for_each(|(index, value)| {
+                // offset calculation by RING_BUFFER_SIZE so we cant negative values
+                let position = (RING_BUFFER_SIZE + write_index - len + index) % RING_BUFFER_SIZE;
+                *value = self.buffer[position].load(Ordering::Relaxed);
+            });
+    }
 
-        // order the ring buffer, and pad with silence
-        let data_chunks = self.buffer[self.index as usize..]
-            .iter()
-            .chain(self.buffer[..self.index as usize].iter())
-            .rev()
-            .chain(std::iter::repeat(&silence));
+    // to simply share tests with the unsafe version
+    #[cfg(test)]
+    fn raw(&self) -> Vec<f32> {
+        let mut slice = vec![0.; RING_BUFFER_SIZE];
 
-        // split the output buffer in same sized chunks
-        let true_size = fft_size.min(buffer.len());
-        let buf_chunks = buffer[0..true_size].chunks_mut(RENDER_QUANTUM_SIZE).rev();
+        self.buffer.iter().zip(slice.iter_mut()).for_each(|(a, b)| {
+            *b = a.load(Ordering::SeqCst);
+        });
 
-        // copy data from internal buffer to output buffer
-        buf_chunks
-            .zip(data_chunks)
-            .for_each(|(b, d)| b.copy_from_slice(&d[..b.len()]));
+        slice
     }
 }
 
-/// Analyser kernel for time domain and frequency data
+// As the analyser is wrapped into a Arc<RwLock<T>> by the analyser node to get interior
+// mutability and expose an immutable public API, we should be ok with thread safety.
 pub(crate) struct Analyser {
-    time: TimeAnalyser,
-
-    fft_planner: RealFftPlanner<f32>,
+    ring_buffer: AnalyserRingBuffer,
+    fft_size: usize,
+    smoothing_time_constant: f64,
+    min_decibels: f64,
+    max_decibels: f64,
+    // If not wrapped into Arc<Mutex<T>>, compiler complains about thread safety:
+    // > `(dyn rustfft::avx::avx_planner::AvxPlannerInternalAPI<f32> + 'static)`
+    // > cannot be shared between threads safely
+    // But the Mutex is ok here as `Analyser` lives outside the audio thread.
```

Comment:
Are you sure this is necessary? It works on my hardware without the Arc+Mutex.
It would also surprise me, Mutex does not magically make non-Send types Send. Being not-Send means you cannot leave the thread you were created on. Arc+Mutex specifically allows one to transfer objects from one thread to another

Reply:
I took the liberty to remove them. Let me know if building fails on your hardware

Reply:
Hah, it fails on CI. The Mutex is required to make the thing `Sync`